src/protocol/templates/ping1d.cpp.in

```diff
@@ -0,0 +1,212 @@
+/*
+* ping1d.cpp
+* A device API for the Blue Robotics Ping1D echosounder
+*
+*~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!
+* THIS IS AN AUTOGENERATED FILE
+* DO NOT EDIT
+*~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!~!
+*
+*/
+
+#include "ping1d.h"
+
+Ping1D::Ping1D(Stream& ser) : _stream ( ser ) {}
+
+Ping1D::~Ping1D()
+{
+{% for msg in messages["ping1D"]["get"] %}
+{% for field in messages["ping1D"]["get"][msg].payload %}
+{% if is_vector(field.type) %}
+    if (_{{field.name}}) {
+       free(_{{field.name}});
+    }
+{% endif %}
+{% endfor %}
+{% endfor %}
+}
+
+PingMessage* Ping1D::read()
+{
+    while(_stream.available()) {
+        if (_parser.parseByte(_stream.read()) == PingParser::NEW_MESSAGE) {
+          handleMessage(&_parser.rxMsg);
+          return &_parser.rxMsg;
+        }
+    }
+    return nullptr;
+}
+
+size_t Ping1D::write(uint8_t* data, uint16_t length)
+{
+    return _stream.write(data, length);
+}
+
+bool Ping1D::initialize(uint16_t ping_interval_ms)
+{
+    if(!request(Ping1DNamespace::Device_id)) {
+        return false;
+    }
+
+    if (!request(Ping1DNamespace::Firmware_version)) {
+        return false;
+    }
+
+    if(!request(Ping1DNamespace::Voltage_5)) {
+        return false;
+    }
+
+    if (!request(Ping1DNamespace::Processor_temperature)) {
+        return false;
+    }
+
+    // Configure ping interval
+    if (!set_ping_interval(ping_interval_ms)) {
+        return false;
+    }
+
+    return true;
+}
+
+PingMessage* Ping1D::waitMessage(enum Ping1DNamespace::msg_ping1D_id id, uint16_t timeout_ms)
+{
+    uint32_t tstart = millis();
+    while (millis() < tstart + timeout_ms) {
+
+        PingMessage* pmsg = read();
+
+        if (!pmsg) {
+            continue;
+        }
+
+        handleMessage(pmsg);
+
+        if (pmsg->message_id() == Ping1DNamespace::Nack) {
+            ping_msg_ping1D_nack nack(*pmsg);
+
+            // TODO remove, this won't ever be true in the current design
```

Comment:
why this message should be removed ? Should be removed now in this PR ?
This check makes sense to me.

Reply:
We are waiting for the response msg id #. The nack id # will match the request id #. Therefore, nack id # will never == the response msg id #, which is what this check is doing.

Reply:
I want to leave it as we need to think about nack, but I don't have a good solution at the time
I can comment the acutal code.

Reply:
We can receive two nacks of different nacked_ids, I don't see why not.
like:
Send something wrong A, sendo something wrong B.
We wait for nack A, that can fail the checksum test and will result in a nullptr.
After that, we can receive nack B, that will be a nack message with different nacked_id value.